### PR TITLE
Deduce CUDA toolkit path from `NVHPC_ROOT` env var

### PR DESCRIPTION
**Pull request by carstenbauer**
First of all, if `NVHPC_ROOT` isn't set, this PR doesn't have any effect.

If `NVHPC_ROOT` is set, the strategy is to look for version-like named subdirectories in `$NVHPC_ROOT/cuda`, which seems to be a canonical path. If we can find a unique one (should be the case), we use it. If we can't, we consider the path that `NVHPC_ROOT` points to itself.

If besides `NVHPC_ROOT`, the `version` preferences is set for `CUDA_Runtime_jll` (say to X.Y) we try to use this information to immediately check if `$NVHPC_ROOT/cuda/X.Y` is available. If it isn't, we fall back to the general "parsing approach" above.

I've tested it on two different clusters and it seems to work fine.

Closes [CUDA.jl#2433](https://github.com/JuliaGPU/CUDA.jl/issues/2433)

### CODE REVIEW

File: src/CUDA_Runtime_Discovery.jl [hunk 1]
```diff
@@ -265,6 +265,41 @@ function has_version_like_name(path)
     return true
 end
 
+function get_version_preference()
+    prefs = get(Base.get_preferences(), "CUDA_Runtime_jll", nothing)
```
Comment (maleadt) [inline src/CUDA_Runtime_Discovery.jl:269]: We can just depend on Preferences.jl here.

Comment (carstenbauer) [reply src/CUDA_Runtime_Discovery.jl:269]: What's the advantage?

Comment (carstenbauer) [reply src/CUDA_Runtime_Discovery.jl:269]: That it's not internal?

Comment (carstenbauer) [reply src/CUDA_Runtime_Discovery.jl:269]: And how, actually? I've never use Preferences.jl to check preferences of a different package. My first attempts failed:

```julia-repl
julia> get(Base.get_preferences(), "CUDA_Runtime_jll", nothing)
Dict{String, Any} with 2 entries:
  "local"   => "true"
  "version" => "12.1"

julia> has_preference(UUID("76a88914-d11a-5bdc-97e0-2f5a05c973a2"), "version")
false

julia> has_preference("CUDA_Runtime_jll", "version")
ERROR: ArgumentError: Cannot resolve package 'CUDA_Runtime_jll' in load path; have you added the package as a top-level dependency?
Stacktrace:
 [1] package_lookup_error(name::String)
   @ Preferences /projects/julia/.julia/packages/Preferences/enLcN/src/utils.jl:88
 [2] has_preference(name::String, key::String)
   @ Preferences /projects/julia/.julia/packages/Preferences/enLcN/src/Preferences.jl:76
 [3] top-level scope
   @ REPL[12]:1
```

Comment (maleadt) [reply src/CUDA_Runtime_Discovery.jl:269]: https://github.com/JuliaGPU/CUDA.jl/blob/bbe625bbd92cf2c3a5fde6aec1c940f0a1e2b039/src/utilities.jl#L93

Comment (maleadt) [reply src/CUDA_Runtime_Discovery.jl:269]: And yeah, I'd rather rely on the package that's intended for this use case, than the internal Base API that could change at any point.

Comment (carstenbauer) [reply src/CUDA_Runtime_Discovery.jl:269]: > [JuliaGPU/CUDA.jl@`bbe625b`/src/utilities.jl#L93](https://github.com/JuliaGPU/CUDA.jl/blob/bbe625bbd92cf2c3a5fde6aec1c940f0a1e2b039/src/utilities.jl#L93)

But the `CUDA_Runtime_jll` module isn't available here (in `CUDA_Runtime_Discovery.jl`). Should we also add `CUDA_Runtime_jll` as a dependency?

Comment (maleadt) [reply src/CUDA_Runtime_Discovery.jl:269]: Maybe hard-code the UUID then and use https://github.com/JuliaPackaging/Preferences.jl/blob/209a45060b8eef13a790300189dfd30c2cb6a1e1/src/Preferences.jl#L28-L35?

Comment (carstenbauer) [reply src/CUDA_Runtime_Discovery.jl:269]: Well, that's what I tried above and it only works if the package is on `LOAD_PATH`:

Not on `LOAD_PATH`, i.e. not in `Project.toml` here:

```julia-repl
julia> Base.get_preferences()
Dict{String, Any} with 2 entries:
  "MPIPreferences"   => Dict{String, Any}("abi"=>"OpenMPI", "_format"=>"1.0", "libmpi"=>"libmpi", "preloads"=>Union{}[], "mp…
  "CUDA_Runtime_jll" => Dict{String, Any}("local"=>"true", "version"=>"12.1")

julia> Base.get_preferences(UUID("76a88914-d11a-5bdc-97e0-2f5a05c973a2")) # UUID of CUDA_Runtime_jll
Dict{String, Any}()
```

With `CUDA_Runtime_jll` in `Project.toml`:

```julia-repl
julia> Base.get_preferences(UUID("76a88914-d11a-5bdc-97e0-2f5a05c973a2"))
Dict{String, Any} with 2 entries:
  "local"   => "true"
  "version" => "12.1"
```

Comment (carstenbauer) [reply src/CUDA_Runtime_Discovery.jl:269]: Note that the docstring of `load_preferences` also says
> loading preferences from all environments that list the given UUID as a direct dependency.

Comment (carstenbauer) [reply src/CUDA_Runtime_Discovery.jl:269]: The way I see it, we have two options:
1) Live with relying on internals (`Base.get_preferences`)
2) Drop the entire preferences story and only use the "parsing approach".

Comment (maleadt) [reply src/CUDA_Runtime_Discovery.jl:269]: Ah, I see. Let's just stick with the internal API then, for simplicity.

**Pull request merged**